Feature/admin role

bryan-robitaille opened the pull request:
# Summary | Résumé

This PR includes enabling the check for the admin role, user administration screen,  and a larger refactor of the middleware functions that encompass the API routes.

**Enabling Admin Role**
The admin boolean field in the user's table is now checked before allowing authenticated access to administrative areas of the application

**User Admin Interface** 
Allows for the setting of administrative rights for users that have attempted to login using the G-Suite CDS account.
![Screen Shot 2022-02-24 at 12 12 02 PM](https://user-images.githubusercontent.com/25329319/155719561-aa9591c5-dd4c-449b-a689-79453576cf87.png)

**Middleware Refactor**

> Exported from `@lib/middleware`

*Middleware Layers*
- validBearerToken: checks for validity of bearer token and adds the formID to the middleware return props
- cors: checks that a specific http request method is allowed and also add's CORS check for browser side calls.
- sessionExists: checks if the user has an active session and if that user is an administrator
- validTemporaryToken: checks for a valid temporary token and adds the email and token to the middleware return props
- jsonValidator: checks for valid JSON form Schema

*Middleware manager*
- middleware: accepts an array of middleware layers, iterates through the layers in order, and consolidates their return props to be passed to the api handler.  The middleware function implements 'fail fast' so that the first middleware failure returns a response and breaks out of any further processing.

# Test instructions | Instructions pour tester la modification

There is a migration in place that will enable all existing users in the DB to admins (no change to existing security controls as these users can already log in).  Default for new users being created is still non-admin status.
- Create a new user in the Database (manual intervention required with pgAdmin) after running migrations. (`INSERT INTO users (name, email) VALUES ('test account 1', 'test@cds-snc.ca')`)
- Login to the Admin pages and view the Users tab. Clicking enable/disable beside the new user should change the admin boolean for the user in the DB.
- Disable the admin privileges on your current user.
- You should be redirected to the Unauthorized page.
- Connect to the Database with you tool of choice and enable admin on your user with (`UPDATE users SET admin=true WHERE id=/your user id/`)  

# Unresolved questions / Out of scope | Questions non résolues ou hors sujet

Logic to set at least one user to Administrator by default will be handled in a future PR that updates Next-Auth to Version 4.  There are breaking changes that need to be addressed, such as the implementation of an ORM.

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:

- [x] Have you completely tested the functionality of change introduced in this PR? Is the PR solving the problem it's meant to solve within the scope of the related issue?
- [x] The PR does not introduce any new issues such as failed tests, console warnings or new bugs.
- [x] If this PR adds a package have you ensured its licensed correctly and does not add additional security issues?
- [x] Is the code clean, readable and maintainable? Is it easy to understand and comprehend.
- [x] Does your code have adequate comprehensible comments? Do new functions have [docstrings](https://tsdoc.org/)?
- [x] Have you modified the change log and updated any relevant documentation?
- [x] Is there adequate test coverage? Both unit tests and end-to-end tests where applicable?
- [x] If your PR is touching any UI is it accessible? Have you tested it with a screen reader? Have you tested it with automated testing tools such as axe?

jeberhardt commented on tests/api/id/[form]/bearer.test.js:
```diff
@@ -4,6 +4,7 @@ import retrieve from "../../../../pages/api/id/[form]/bearer";
 import executeQuery from "../../../../lib/integration/queryManager";
 import jwt from "jsonwebtoken";
 import { logMessage } from "../../../../lib/logger";
+import { checkLogs } from "../../../../lib/jestUtils";
```
It would be cleaner if it was `@lib/jestUtils`. I believe the issue that required all relative paths was fixed a while ago.

bryan-robitaille replied:
Thanks James, I'll go through all the tests and refactor since I'm in there.

jeberhardt commented on public/static/locales/fr/admin-users.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "title": "Manage Users",
```
Most of these are still in English.

bryan-robitaille replied:
Will do the rough translations.

bryan-robitaille replied:
Done

jeberhardt commented on lib/users.ts:
```diff
@@ -0,0 +1,35 @@
+ * @returns An array of all Users
+ */
+export const getUsers = async (): Promise<AuthenticatedUser[]> => {
+  const result = await executeQuery(
```
Should there be a try/catch around this block of code?

bryan-robitaille replied:
Yes there should.  Thanks for catching.  I hadn't done a complete recheck of everything yet in the PR, hence the draft PR status.

jeberhardt commented on lib/tests/httpSessionExist.test.js:
```diff
@@ -1,35 +1,36 @@
 import { createMocks } from "node-mocks-http";
 import client from "next-auth/client";
-import isUserSessionExist from "../middleware/HttpSessionExist";
+import isUserSessionExist from "../middleware/HttpSessionExists";
```
Lowercase `H` on the filename: `httpSessionExists`.

jeberhardt commented on .vscode/settings.json:
```diff
@@ -6,4 +6,5 @@
     "storybook-static/static/locales"
   ],
   "jest.jestCommandLine": "yarn test",
+  "i18n-ally.keystyle": "nested",
```
We might want to ignore this file from the git repo.

bryan-robitaille replied:
Agreed, now that we have the basic settings already tracked in the file for prettier we can untrack it.

jeberhardt commented on pages/admin/users.tsx:
```diff
@@ -0,0 +1,112 @@
+}) => {
+  const { refreshData, isRefreshing } = useRefresh([isAdmin]);
+  return (
+    <Button
```
I think this is missing the `ariaProps` that you've added to the `<Button>` component above.

bryan-robitaille replied:
In this case the Aria-props isn't necessary.  I've simply added the feature for future use.
Instead of using aria-props and creating a new element, in order to reference it's id, I've used a span within the button that is only screen-reader visible.

jeberhardt commented on lib/auth.ts:
```diff
@@ -25,6 +32,15 @@ export function requireAuthentication(
+    if (!session.user?.admin) {
+      return {
+        redirect: {
+          destination: `/${context.locale}/welcome-bienvenue/`,
```
Should this redirect to an error page, giving them appropriate information on why they didn't go to the page they wanted to see? Bouncing them over the welcome page may be confusing if they think they are legitimately supposed to be allowed to see the originally requested page.

bryan-robitaille replied:
I'll add in a temporary error page for 'not authorized - must be admin'.

jeberhardt commented on lib/jestUtils.ts:
```diff
@@ -0,0 +1,18 @@
+  if (foundLogs.length > 0) {
+    return true;
+  }
+  return false;
```
Could be simplified: `return foundLogs.length > 0`

bryan-robitaille replied:
Definitely, good catch.

jeberhardt commented on components/globals/AdminNav.tsx:
```diff
@@ -21,6 +21,9 @@ const AdminNav = (props: AdminNavProps): React.ReactElement => {
           <Link href="/admin/vault">Form Vault</Link>
         </li>
+        <li className="gc-horizontal-item">
+          <Link href="/admin/users">Users</Link>
```
Noticing that we don't have any translations for the admin page.

bryan-robitaille replied:
When the page was originally built as a concept, and given the scope was purely internal at the time, we didn't do the translations.  I can add them quickly in this PR for the menu items.

jeberhardt replied:
Or just create a new card so that it's on the list of things to do. No need to delay this PR.

bryan-robitaille replied:
Added translations

jeberhardt commented:
Seem to be missing tests and stories for `users.tsx`.

Moro-Code commented on pages/api/id/[form]/bearer.ts:
```diff
@@ -111,4 +110,4 @@ export const getTokenById = async (formID: string): Promise<QueryResult> => {
 };
 
-export default isRequestAllowed(["GET", "POST"], isUserSessionExist(handler));
+export default middleware([httpMethodAllowed(["GET", "POST"]), sessionExists()], handler);
```
Definitely much much cleaner :)

Moro-Code commented on pages/api/id/[form]/bearer.ts:
```diff
@@ -14,14 +13,14 @@ const handler = async (req: NextApiRequest, res: NextApiResponse): Promise<void>
       return await getToken(req, res);
     } catch (err) {
-      logMessage.error(err);
+      logMessage.error(err as Error);
```
Curious is there really a need to typecast here ? Should it not already be inferred to be of type Error considering its in a catch block

bryan-robitaille replied:
On my IDE the linter continually complains and this is because the catch block cannot be inferred as a type Error.  It is always inferred as type unknown. I don't know why the other team members linters don't complain about this?
https://stackoverflow.com/questions/69021040/why-catch-clause-variable-type-annotation-must-be-any

Moro-Code replied:
Do not get the same complaints on my end. If it's in a catch block it should be inferred that it's going to be of type Error. The link you sent seems to be someone trying to typecast a specific error type

bryan-robitaille replied:
In newer versions of typescript it's only ever inferred as `unknown`
https://kentcdodds.com/blog/get-a-catch-block-error-message-with-typescript

I can modify the `logMessage` handler instead to do strong type checking internally however this seemed more readable at the time.

bryan-robitaille replied:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#unknown-on-catch-clause-bindings

Moro-Code commented on pages/api/id/[form]/bearer.ts:
```diff
@@ -14,14 +13,14 @@ const handler = async (req: NextApiRequest, res: NextApiResponse): Promise<void>
       return await createToken(req, res);
     } catch (err) {
-      logMessage.error(err);
+      logMessage.error(err as Error);
```
Same here

Moro-Code commented on pages/api/id/[form]/owners.ts:
```diff
@@ -104,7 +104,7 @@ export async function addEmailToForm(req: NextApiRequest, res: NextApiResponse):
     return res.status(200).json({ success: result.rows[0] });
   } catch (error) {
-    const message = error.message;
+    const message = (error as Error).message;
```
unnecessary typecasting ?

bryan-robitaille replied:
See other comment about `error` in a catch block always being typed `unknown`

Moro-Code commented on lib/middleware/cors.ts:
```diff
@@ -0,0 +1,29 @@
+import { NextApiRequest, NextApiResponse } from "next";
```
Not seeing any tests for this file.

bryan-robitaille replied:
Tests created

Moro-Code commented on pages/api/retrieval.ts:
```diff
@@ -67,7 +67,7 @@ async function getFormResponses(
     formResponses = await db.send(new QueryCommand(getItemsDbParams));
   } catch (error) {
-    logMessage.error(error);
+    logMessage.error(error as Error);
```
unnecessary typecasting ?

Moro-Code commented on pages/api/retrieval.ts:
```diff
@@ -116,10 +116,10 @@ async function getFormResponses(
     //print submission ids as string
     logMessage.warn(submissionIDlist ? submissionIDlist.toString() : "Empty submissionID list");
-    logMessage.error(error);
+    logMessage.error(error as Error);
```
unnecessary typecasting ?

Moro-Code commented on pages/api/retrieval.ts:
```diff
@@ -29,14 +30,13 @@ import checkIfValidTemporaryToken from "@lib/middleware/httpRequestHasValidTempT
+  const { maxRecords = "10", formID } = req.query;
   //Check maxRecords aren't repeated
-  if (Array.isArray(maxRecords) || !formID) return res.status(400).json({ error: "Bad Request" });
+  if (Array.isArray(maxRecords) || Array.isArray(formID) || !formID)
```
If it's going to wrap around anyways might as well add brackets to avoid confusion. Good catch that formID can also be an array

bryan-robitaille replied:
:+1

Moro-Code commented on pages/api/retrieval.ts:
```diff
@@ -116,10 +116,10 @@ async function getFormResponses(
 
 // Only a GET request is allowed
-export default isRequestAllowed(["GET"], checkIfValidTemporaryToken(getFormResponses));
+export default middleware([httpMethodAllowed(["GET"]), validTemporaryToken()], getFormResponses);
```
Love seeing this new form. Much more readable!

Moro-Code commented on pages/api/users.ts:
```diff
@@ -0,0 +1,38 @@
+import { NextApiRequest, NextApiResponse } from "next";
```
New file needs tests

bryan-robitaille replied:
done

Moro-Code commented on pages/api/users.ts:
```diff
@@ -0,0 +1,38 @@
+  res.status(200).json({ users });
+};
+
+const postMethod = async (req: NextApiRequest, res: NextApiResponse) => {
```
This seems like a PUT to me not a POST. It's an idempotent request.

Moro-Code commented on styles/app.scss:
```diff
@@ -9,4 +9,6 @@
 
-@import "~tailwindcss/utilities";
\ No newline at end of file
+// Tailwind Utilities needs to be loaded last to use inline
```
Thanks for the comment !

Moro-Code commented on tests/api/id/[form]/bearer.test.js:
```diff
@@ -1,14 +1,15 @@
 import client from "next-auth/client";
 import retrieve from "../../../../pages/api/id/[form]/bearer";
-import executeQuery from "../../../../lib/integration/queryManager";
+import executeQuery from "@lib/integration/queryManager";
```
That is definitely way better

Moro-Code commented on tests/api/id/[form]/bearer.test.js:
```diff
@@ -252,15 +255,17 @@ describe("/id/[form]/bearer", () => {
-      expect(logMessageDebugSpy.mock.calls[0][0]).toBe(
-        "A bearer token was attempted to be created for form 1 by user Admin user but the form does not exist"
-      );
+      expect(
```
Is this really more efficient? This takes 6 lines as opposed to the previous 4 and at least to me the first option is more readable

bryan-robitaille replied:
It's not about efficiency but that the original statement checked an exact position in the array.  If in future code changes a `logMessage.debug()` is added somewhere before this checked call, like in an additional middleware, this check will fail.  The `checkLogs` function instead looks for the exact expected message in the logMessage array which removes the importance of knowing its exact position.

Moro-Code commented on tests/templates.test.js:
```diff
@@ -92,14 +92,13 @@ describe("Test JSON validation scenarios", () => {
     client.getSession.mockReturnValueOnce([mockSession, false]);
     const { req, res } = createMocks({
-      method: "POST",
+      method: "INSERT",
```
I don't think this is correct. This is the method of the mocked HTTP request

bryan-robitaille replied:
Good catch.

bryan-robitaille replied:
This was a cross change that got captured in a stash.  Reverting

Moro-Code commented on pages/api/auth/[...nextauth].js:
```diff
@@ -15,13 +15,29 @@ export default NextAuth({
-      })
-    : null,
+      };
+      const extendedSession = { ...session, ...extendedInfo };
```
Could you just do `return { ...session, ...extendedInfo }` ?

bryan-robitaille replied:
yes, yes I could.  Was originally nicer to debug and step through this way.

bryan-robitaille commented on pages/admin/users.tsx:
```diff
@@ -0,0 +1,112 @@
+import { serverSideTranslations } from "next-i18next/serverSideTranslations";
```
I can write tests covering the functionality.

bryan-robitaille replied:
Using cypress for this won't be possible at the moment due to the auth requirement.  Will address in PR when we implmeent Next Auth V4.

bryan-robitaille commented on lib/middleware/httpMethodAllowed.ts:
```diff
@@ -0,0 +1,33 @@
+
+      // Also check CORS by evaluating the origin of the request
+      // TODO: Modify origin to acceptable parameter
+      const blockedByCors = await cors({ origin: "*", methods: methods })(req, res);
```
My assumption was that when the httpCheckMethod middleware is used it would also always be protected by CORS.  While we are already checking for the http method the inclusion of origin might something we want to include later on for requests originating from a browser.  It can be easily ripped out if we don't want to have any CORS checks for browserside calls.

Moro-Code replied:
Alright no worries we can leave it in. Was just wondering if it would be better to leave it out so as to follow a single responsibility pattern for middlewares

bryan-robitaille commented on lib/middleware/middleware.ts:
```diff
@@ -0,0 +1,35 @@
+      for (const middlewareLayer of middlewareArray) {
+        const { pass, props: middlewareProps } = await middlewareLayer(req, res, props);
+        if (!pass) {
+          blockedByMiddleware = true;
```
Sure, will refactor.

bryan-robitaille commented:
Ready for another pass at review.
Major changes:
- Added migration to make existing users administrators.  All new users default to admin=false (existing functionality).
- Merged CORS and HTTPmethodAllowed into a more cohesive module.  CORS check first / if it doesn't originate from a browser then manual check on the http method.
- Added tests for users API, cors middleware, and middleware
- Added missing translations to admin nav, and manage users page

craigzour reviewed (dismissed):
(oops I forgot to select `request changes`)

Amazing work @bryan-robitaille :) Tested it locally and did not find any issue.

Just a little change to bring in the French translation file.

I love the middleware refactoring, it reminded me a feature I worked on a few years ago. We had to create a connection mechanism (between two devices in Bluetooth) with a few checks that were running sequentially in order to approve the connection.

Not gonna lie but reviewing this PR was tough because there is a lot of changes 😄 One of the reason why I always liked to reorganize my commits at the end of my work so that they tackle only one problem at a time and they can be reviewed one by one with some context.

bryan-robitaille commented:
> One of the reason why I always liked to reorganize my commits at the end of my work so that they tackle only one problem at a time and they can be reviewed one by one with some context.

Thanks @craigzour !  I hadn't though about that.  I'll definitely try to do that in the future.
